runtime/lua/vim/lsp/util.lua

```diff
@@ -1878,6 +1888,40 @@ function M.lookup_section(settings, section)
   return settings
 end
 
+
+--- Convert diagnostics grouped by bufnr to a list of items for use in the
+--- quickfix or location list.
+---
+--@param diagnostics_by_bufnr table bufnr -> Diagnostic[]
+--@param predicate an optional function to filter the diagnostics.
+--                  If present, only diagnostic items matching will be included.
+--@return table (A list of items)
+function M.diagnostics_to_items(diagnostics_by_bufnr, predicate)
+  local items = {}
+  for bufnr, diagnostics in pairs(diagnostics_by_bufnr or {}) do
+    for _, d in pairs(diagnostics) do
+      if not predicate or predicate(d) then
+        table.insert(items, {
+          bufnr = bufnr,
+          lnum = d.range.start.line + 1,
+          col = d.range.start.character + 1,
```

Comment:
I left out the `character_offset` call because it is wrong. We'd need to convert utf-16 offsets to byte positions if we want to be correct here.